benchmark/single-source/KeyPathPerformanceTests.swift

```diff
@@ -0,0 +1,1009 @@
+//===--- KeyPathPerformanceTests.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import TestsUtils
+
+public let benchmarks = [
+  BenchmarkInfo(
+    name: "KeyPathNestedStructs",
+    runFunction: run_KeyPathNestedStructs,
+    tags: [.validation, .api],
+    setUpFunction: setupKeyPathNestedStructs
+  ),
+  BenchmarkInfo(
+    name: "KeyPathDirectAccess",
+    runFunction: run_testDirectAccess,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+  BenchmarkInfo(
+    name: "KeyPathReadPerformance",
+    runFunction: run_testKeyPathReadPerformance,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+  BenchmarkInfo(
+    name: "KeyPathWritePerformance",
+    runFunction: run_testKeyPathWritePerformance,
+    tags: [.validation, .api],
+    setUpFunction: setupSingleton
+  ),
+]
```

Comment:
It would also be very useful to add benchmarks for other keypath variants, like getters+setters.

Reply:
I'm wondering what is meant by this: does this mean adding benchmarks featuring ReferenceWritableKeyPaths, PartialKeyPaths, and others? 
Or, benchmarks for KeyPath syntax like this?
`get { ... #keyPath(...) } `

Reply:
By the way, this benchmark is already getting somewhat long (over 1300 lines). Would it be a good candidate for `multi-source`?

Reply:
Example for a gettable property:
```
struct S {
  var x: Int { return 27 }
}

let kp = \S.x
```

> Would it be a good candidate for multi-source?

No, keeping it in a single file is fine.

Reply:
There are several kind of keypath components:
* stored property (what you are testing)
* gettable property (my example)
* settable property (the same, just for writing)
* tuple element
* optional chain
* optional force
* optional wrap

Those are handled differently in the runtime. Therefore it would be nice to have benchmarks for all of the keypath component types + combinations of them, e.g.

```
struct T {
  var y: Int
}

struct S {
  var x: (Int, T?) { return (27, T(y: 42)) }
}

let kp = \S.x.1?.y
```

Reply:
Ah yes, I see that they're handled differently, especially inside `_projectReadOnly<CurValue, NewValue, LeafValue>` in `KeyPath.swift`.

Reply:
I've added benchmarks that'll test the performance of each type of KeyPathComponent. There is now a MarkDown table near the top of the file that indicates which benchmark stresses which type of KeyPathComponent.
I did find another error in the CI output indicating that the setup time for `KeyPathWritePerformance` was too high; I've moved its KeyPaths, and the ones for `KeyPathReadPerformance`, into the singleton.

Reply:
I've updated the main message associated with the PR to discuss the benchmarks that have been added since then.
What is the convention for resolving conversations (via the `Resolve conversation` button) during workflows like this?

Reply:
As you like. Just let me somehow know when the PR is ready for another round of review.

Reply:
Sound good!